src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -29,16 +29,84 @@ public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchA
+
+            // Get the metadata for the latest version of the installed packages and transitive packages
+            IEnumerable<PackageCollectionItem> installedPackagesLatest = GetLatestPackages(_installedPackages);
+            IEnumerable<PackageCollectionItem> latestTransitivePackages = GetLatestPackages(transitivePkgsWithOrigins);
+
+            var installedItems = await GetMetadataForPackagesAsync(PerformLookup(installedPackagesLatest, searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
+            var transitiveItems = await GetMetadataForPackagesAsync(PerformLookup(latestTransitivePackages, searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
 
-            IPackageSearchMetadata[] installedItems = await GetMetadataForPackagesAndSortAsync(PerformLookup(_installedPackages.GetLatest(), searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
-            IPackageSearchMetadata[] transitiveItems = await GetMetadataForPackagesAndSortAsync(PerformLookup(transitivePkgsWithOrigins.GetLatest(), searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
-            IPackageSearchMetadata[] items = installedItems.Concat(transitiveItems).ToArray();
+            // Get metadata from identity for the rest of the installed and transitive packages
+            var remainingInstalledItems = await GetRemainingPackagesMetadataAsync(_installedPackages, searchToken, cancellationToken);
+            var remainingTransitiveItems = await GetRemainingPackagesMetadataAsync(transitivePkgsWithOrigins, searchToken, cancellationToken);
+
+            // Combine the installed and transitive packages
+            IPackageSearchMetadata[] items = installedItems
+                .Concat(remainingInstalledItems)
+                .Concat(transitiveItems)
+                .Concat(remainingTransitiveItems)
+                .ToArray();
 
             return CreateResult(items);
         }
 
+        private static IEnumerable<PackageCollectionItem> GetTransitivePackagesWithOrigins(IEnumerable<PackageCollectionItem> transitivePackages)
+        {
+            return transitivePackages
+                .Where(t => t.PackageReferences.All(x => x is ITransitivePackageReferenceContextInfo y && y.TransitiveOrigins.Any()));
+        }
+
+        private static IEnumerable<PackageCollectionItem> GetLatestPackages(IEnumerable<PackageCollectionItem> packages)
+        {
+            return packages
+                .GroupById()
+                .Select(g => g.OrderByDescending(x => x.Version).First());
+        }
+
+        private static async Task<IPackageSearchMetadata[]> GetRemainingPackagesMetadataAsync(IEnumerable<PackageCollectionItem> packages, FeedSearchContinuationToken searchToken, CancellationToken cancellationToken)
+        {
+            var remainingPackages = packages
```

Comment:
I believe the installed packages are already grouped by ID here: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Common/PackageCollection.cs#L64

Reply:
The grouping you referenced is grouping by PackageIdentity which is an object that consists of the Id (name of package) and Version. Here, we are grouping by the more specific Id only, producing a list of versions for the grouped Id.

Reply:
Grouping by PackageIdentity will group on PackageId + PackageVersion.

Here we are grouping on PackageId only so we can get a list of PackageVersions grouped by PackageId. The Package Id is the name of the package.

As an example if we have Newtonsoft.Json versions 13.0.0 and 12.0.0 installed across the solution, there will be two package references, and so we'll have two PackageIdentity objects like this:
PackageIdentity { Id = "Newtonsoft.Json", Version = "13.0.0" }
PackageIdentity { Id = "Newtonsoft.Json", Version = "12.0.0" }

When we group by Id here, we will group all Newtonsoft.Json Ids and have a list of the Versions { "13.0.0", "12.0.0" } so they're sorted, so we can skip the first one per each package and process the rest.